### ISSUE TITLE
Istio galley failing liveness probes

### ISSUE THREAD

**Issue by dmyerscough** [istio/istio#7586]
**Describe the bug**

The istio-galley pod fails the Liveness probe which causes the container to be constantly restarted.

**Expected behavior**

I expect the istio-galley not to constantly fail

**Steps to reproduce the bug**

Install Istio using the helm template. 

**Version**
```bash
$ kubectl get nodes
NAME                       STATUS    ROLES     AGE       VERSION
aks-agentpool-36280988-0   Ready     agent     2h        v1.10.6
aks-agentpool-36280988-1   Ready     agent     2h        v1.10.6
aks-agentpool-36280988-2   Ready     agent     2h        v1.10.6

$ istioctl version
Version: 1.0.0
GitRevision: 3a136c90ec5e308f236e0d7ebb5c4c5e405217f4
User: root@71a9470ea93c
Hub: gcr.io/istio-release
GolangVersion: go1.10.1
BuildStatus: Clean
```

**Is Istio Auth enabled or not?**

```bash
$ helm template install/kubernetes/helm/istio --name istio --namespace istio-system > $HOME/istio.yaml
$ kubectl create -f $HOME/istio.yaml
```

**Environment**
Azure

**Cluster state**

[istio-dump.tar.gz](https://github.com/istio/istio/files/2252232/istio-dump.tar.gz)

**Comment by vadimeisenbergibm** [comment]
May be related to https://github.com/istio/istio/issues/7593.

**Comment by vadimeisenbergibm** [comment]
May be the same problem as in https://github.com/istio/istio/issues/7586

**Comment by ayj** [comment]
Readiness and liveliness probes were failing on some providers because GET requests were blocking for multiple seconds. Mitigate the issue by decreasing the frequency of GET (to avoid possible throttling) and increase the acceptable health check interval from 4s to 10s.

See https://github.com/istio/istio/pull/7605 for short term fix. Longer term fix requires switching to proper controller-style reconciliation. That work should be aligned with the sidecar injector.

**Comment by dmyerscough** [comment]
@ayj I reinstalled Istio with the settings from #7605 and the pod isn't in a CrashLoop now. However, the pod still restarts with the following message:-

```
  Warning  Unhealthy  6m (x83 over 1h)  kubelet, aks-agentpool-36280988-1  (combined from similar events): Readiness probe failed: fail on inspecting path /health: file /health is too old (last modified time 2018-08-09 05:38:26.005282502 +0000 UTC, should be within 10.01s)
```

Is there any documentation for building `galley/pkg/crd/validation/webhook.go` and applying it to a cluster?

**Comment by ayj** [comment]
https://github.com/istio/istio/wiki/Dev-Guide#using-the-code-base

```bash
(... check out istio.io/istio repo ...)
export HUB=<your docker repo>
export TAG=<your tag>
make galley docker.galley push.docker.galley
kubectl -n istio-system set image deployment/istio-galley validator=${HUB}/galley:${TAG}
```

**Comment by dmyerscough** [comment]
@ayj after a certain period of time; the Istio-galley pod just goes into a Crashloop and causes the service mesh to become unavailable to config updates :(

I tried to get the logs but none are present.

**Comment by ChrisPei** [comment]
Hey guys, what's the test result?

**Comment by olegchorny** [comment]
@ChrisPei I've tried the latest daily build where is some fixes were added already, unfortunately, galley still not stable yet.

**Comment by ayj** [comment]
I had to decouple the health check and reconciliation updates to make this work on AKS (see https://github.com/istio/istio/pull/7986). I ran this overnight and didn't see any pod restarts.

**Comment by ayj** [comment]
@olegchorny, @dmyerscough - the [1.0 release branch](https://github.com/istio/istio/tree/release-1.0) includes the proposed fix for this issue. I tested tested the change overnight but it would be great if someone else that reported the problem could also confirm the issue is resolved.

**Comment by BernhardRode** [comment]
@ayj 
Was this (https://github.com/istio/istio/issues/7675#issuecomment-415305317) the correct way?

If yes, i can confirm the issue is fixed for me.

**Comment by olegchorny** [comment]
@ayj Seems like you've fixed not only galley, but other issues, related to dozens messaged from pilot: 
_"err the server was unable to return a response in the time allotted, but may still be processing the request (get pods)"._ 
Istio was barely operable because of them, but right now it looks much better.

**Comment by polothy** [comment]
Galley appears to be fixed, but problems with pilot and telemetry now.  See https://github.com/istio/istio/issues/7675#issuecomment-415447894 and https://github.com/istio/istio/issues/7675#issuecomment-415452434

**Comment by brbarnett** [comment]
I'm also seeing this on AKS:

    NAME                                        READY     STATUS              RESTARTS   AGE
    grafana-c9f9995bd-2fx7c                     0/1       CrashLoopBackOff    21         1h
    istio-citadel-f89fbffc-mtcw8                0/1       CrashLoopBackOff    21         1h
    istio-egressgateway-5f47f7849c-xtx4l        1/1       Running             13         1h
    istio-galley-5b574674b-4jlwf                0/2       ContainerCreating   0          1h
    istio-grafana-post-install-t26l8            0/1       Completed           0          1h
    istio-ingressgateway-846fbc6cbd-qd57b       1/1       Running             13         1h
    istio-pilot-5cbbd796cc-mdm76                1/2       CrashLoopBackOff    27         1h
    istio-policy-dd789dd5-8llx5                 2/2       Running             13         1h
    istio-sidecar-injector-86df49b8c5-56k6n     0/1       ContainerCreating   0          1h
    istio-statsd-prom-bridge-549d687fd9-mkdqz   1/1       Running             0          1h
    istio-telemetry-56c64b844-lndhr             1/2       CrashLoopBackOff    34         1h
    istio-tracing-776595b9fb-rqmm8              1/1       Running             0          1h
    prometheus-6ffc56584f-25j6x                 1/1       Running             0          1h
    servicegraph-7f4dd8cb46-6lggh               1/1       Running             0          1h

I haven't deployed anything on the cluster other than Istio (commit [0668668](https://github.com/istio/istio/commit/0668668688dcb50cd6626767b9f54437d56f41d2)) via Helm. Installation command:

    helm install install/kubernetes/helm/istio `
        --name istio `
        --namespace istio-system `
        --set global.controlPlaneSecurityEnabled=true `
        --set global.mtls.enabled=true `
        --set grafana.enabled=true `
        --set grafana.persist=true `
        --set servicegraph.enabled=true `
        --set tracing.enabled=true

**Comment by olegchorny** [comment]
```
istio-system   grafana-777bdf8c65-hvff5                    1/1       Running            0          7h
istio-system   istio-citadel-7bcf9c887-kkpdp               1/1       Running            0          7h
istio-system   istio-cleanup-secrets-tk2nh                 0/1       Completed          0          7h
istio-system   istio-egressgateway-ffb7b49b9-ctfqv         1/1       Running            0          7h
istio-system   istio-galley-5545644d94-vcjpt               1/1       Running            0          7h
istio-system   istio-grafana-post-install-ttx5p            0/1       Completed          2          7h
istio-system   istio-ingressgateway-5fcfd54b78-zn2w7       1/1       Running            0          7h
istio-system   istio-pilot-76454c4c7b-vzcxs                2/2       Running            0          7h
istio-system   istio-policy-6cdcf67597-8kmfp               1/2       CrashLoopBackOff   7          10m
istio-system   istio-policy-6cdcf67597-bzcx9               1/2       CrashLoopBackOff   137        6h
istio-system   istio-policy-6cdcf67597-qm7s9               2/2       Running            0          7h
istio-system   istio-policy-6cdcf67597-vm9ch               1/2       CrashLoopBackOff   119        5h
istio-system   istio-policy-6cdcf67597-x77nf               1/2       CrashLoopBackOff   99         4h
istio-system   istio-security-post-install-j59xq           0/1       Completed          2          7h
istio-system   istio-sidecar-injector-567d44fdd6-zrs79     1/1       Running            2          7h
istio-system   istio-statsd-prom-bridge-549d687fd9-bhfp6   1/1       Running            0          7h
istio-system   istio-telemetry-85d747d5fb-l49mk            2/2       Running            16         7h
istio-system   istio-telemetry-85d747d5fb-mr8zv            2/2       Running            0          7h
istio-system   istio-telemetry-85d747d5fb-p8hp9            2/2       Running            19         6h
istio-system   istio-telemetry-85d747d5fb-r7rtq            2/2       Running            51         6h
istio-system   istio-telemetry-85d747d5fb-zh792            1/2       CrashLoopBackOff   141        6h
istio-system   istio-tracing-7596597bd7-9z8dm              1/1       Running            0          7h
istio-system   prometheus-6ffc56584f-n9jk7                 1/1       Running            0          7h
istio-system   servicegraph-b8f946545-82zn2                1/1       Running            0          7h

```
```
Events:
  Type     Reason     Age                 From                               Message
  ----     ------     ----                ----                               -------
  Warning  Unhealthy  13m (x411 over 6h)  kubelet, aks-agentpool-10685398-2  Liveness probe failed: Get http://10.244.0.11:9093/version: dial tcp 10.244.0.11:9093: connect: connection refused
  Warning  BackOff    3m (x1625 over 6h)  kubelet, aks-agentpool-10685398-2  Back-off restarting failed container

**Comment by BernhardRode** [comment]
Sad but true, I can confirm the exact same behaviour:

➜  istio git:(master) k get pods -n istio-system
NAME                                        READY     STATUS             RESTARTS   AGE
grafana-777bdf8c65-ms9s9                    1/1       Running            0          19h
istio-citadel-7bcf9c887-7vd8l               1/1       Running            0          19h
istio-cleanup-secrets-vh964                 0/1       Completed          0          19h
istio-egressgateway-68b6466664-slk2c        1/1       Running            0          19h
istio-galley-5545644d94-4lcx9               1/1       Running            0          19h
istio-grafana-post-install-r2vf6            0/1       Completed          0          19h
istio-ingressgateway-66746dcc69-qbqzx       1/1       Running            0          17h
istio-pilot-67fbdb9ffb-frbbv                2/2       Running            0          19h
istio-policy-67d7f55869-24dst               2/2       Running            0          19h
istio-policy-67d7f55869-79qtg               1/2       CrashLoopBackOff   151        7h
istio-policy-67d7f55869-gwrkb               1/2       CrashLoopBackOff   145        6h
istio-policy-67d7f55869-ntphw               1/2       CrashLoopBackOff   131        6h
istio-policy-67d7f55869-w5x9m               2/2       Running            141        6h
istio-sidecar-injector-567d44fdd6-vxtr4     1/1       Running            0          19h
istio-statsd-prom-bridge-549d687fd9-bcnx6   1/1       Running            0          19h
istio-telemetry-9c9798964-692tb             2/2       Running            149        6h
istio-telemetry-9c9798964-9hsnn             1/2       CrashLoopBackOff   151        7h
istio-telemetry-9c9798964-bvhtg             2/2       Running            4          7h
istio-telemetry-9c9798964-dlks2             2/2       Running            0          19h
istio-telemetry-9c9798964-qrbjx             2/2       Running            0          7h
istio-tracing-7596597bd7-5wrk7              1/1       Running            0          19h
prometheus-6ffc56584f-kxzn9                 1/1       Running            0          19h
servicegraph-b8f946545-f4p29                1/1       Running            2          19h

**Comment by ayj** [comment]
Do you see the policy and telemetry pods in `CrashLoopBackOff` with the latest 1.0 branch snapshot _and_ galley validation disabled?

**Comment by brbarnett** [comment]
Can you confirm you're referring to commit [a66a221](https://github.com/istio/istio/commit/a66a221bfe4cc068abad0c1fba93093600a5c6ef) and `--set global.configValidation=false` flag for Helm?

**Comment by ayj** [comment]
yes

**Comment by brbarnett** [comment]
It appears I still have the same problem:

    NAME                                       READY     STATUS             RESTARTS   AGE
    grafana-8cddb6cf5-fv2c2                    1/1       Running            0          4m
    istio-citadel-75cfbdf5d-ndsqn              1/1       Running            0          4m
    istio-egressgateway-79db7fcbff-c7pcg       1/1       Running            0          4m
    istio-galley-575c744c8d-g5xlg              0/1       CrashLoopBackOff   4          4m
    istio-grafana-post-install-zj8kj           0/1       Completed          0          4m
    istio-ingressgateway-7c67466c75-74tr5      1/1       Running            0          4m
    istio-pilot-7d7b8796b5-jwtjr               2/2       Running            0          4m
    istio-policy-54446f5979-wxtkp              1/2       CrashLoopBackOff   4          4m
    istio-sidecar-injector-fdf6b7ddf-tcb4f     1/1       Running            0          4m
    istio-statsd-prom-bridge-8d8b69b54-77hd4   0/1       CrashLoopBackOff   5          4m
    istio-telemetry-54cf4f45db-g8fl8           1/2       CrashLoopBackOff   5          4m
    istio-tracing-6559954886-6nqtn             1/1       Running            0          4m
    prometheus-657487549-p84cm                 1/1       Running            0          4m
    servicegraph-5b6b5b67c5-6vvpw              1/1       Running            0          4m

Installation command:

    helm install install/kubernetes/helm/istio `
        --name istio `
        --namespace istio-system `
        --set global.controlPlaneSecurityEnabled=true `
        --set global.configValidation=false `
        --set global.mtls.enabled=true `
        --set grafana.enabled=true `
        --set grafana.persist=true `
        --set servicegraph.enabled=true `
        --set tracing.enabled=true

**Comment by dmyerscough** [comment]
From the master branch there was a bunch of containers crashing. However, from the `release-1.0` branch the installation looks good.

@brbarnett did you use the `release-1.0` branch to install Istio?

```bash
$ kubectl -n istio-system get pods
NAME                                        READY     STATUS      RESTARTS   AGE
istio-citadel-66cbc4c899-prnxp              1/1       Running     0          1m
istio-cleanup-secrets-zpssq                 0/1       Completed   0          1m
istio-egressgateway-867699bb4d-md8c8        1/1       Running     0          1m
istio-galley-75679b695b-lhdft               1/1       Running     0          1m
istio-ingressgateway-7f75997c8c-fqf8s       1/1       Running     0          1m
istio-pilot-d548b44f6-crjvk                 2/2       Running     0          1m
istio-policy-5477567644-k6ggg               2/2       Running     0          1m
istio-sidecar-injector-c8c4c568b-smhht      1/1       Running     0          1m
istio-statsd-prom-bridge-7f44bb5ddb-jdz2n   1/1       Running     0          1m
istio-telemetry-79d89d86d5-b77bf            2/2       Running     0          1m
prometheus-84bd4b9796-r77nc                 1/1       Running     0          1m
```

I will leave this running for the next few days to see if Galley goes into a CrashLoop.

**Comment by dmyerscough** [comment]
@ayj the changes you made look to have fixed the issue; I haven't seen any restarts :+1:

**Comment by brbarnett** [comment]
I used the [1.0.0-snapshot.2](https://github.com/istio/istio/releases/tag/1.0.0-snapshot.2) tag. Which commit/tag are you working off of, @dmyerscough?

**Comment by dmyerscough** [comment]
@brbarnett I worked off `release-1.0`

```bash
$ git branch
  master
* release-1.0
```

**Comment by brbarnett** [comment]
The `release-1.0` branch solved my problem. All pods are running and healthy in AKS. I misunderstood their branching strategy and assumed `master` would be stable v1.0. Thanks for your help, @dmyerscough!

**Comment by ayj** [comment]
related https://github.com/Azure/AKS/issues/620

Let's close this issue since the galley specific problem seems to be addressed in the latest 1.0.1 release branch. Let's use https://github.com/istio/istio/issues/7675 to track the telemetry/policy liveness issue.
